resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala

```diff
@@ -265,9 +268,53 @@ private[spark] class ExecutorPodFactoryImpl(
     val executorPodWithNodeAffinity =
         nodeAffinityExecutorPodModifier.addNodeAffinityAnnotationIfUseful(
             executorPodWithInitContainer, nodeToLocalTaskCount)
-    new PodBuilder(executorPodWithNodeAffinity)
+
+    val (executorPodWithTempLocalDirs, executorContainerWithTempLocalDirs) =
+        if (shuffleManager.isEmpty) {
+          // If we're not using the external shuffle manager, we should use emptyDir volumes for
+          // shuffle directories since it's important for disk I/O for these directories to be
+          // performant. If the user has not provided a local directory, instead of using the
+          // Java temporary directory, we create one instead. This is because we want to avoid
+          // as much as possible mounting an emptyDir which overlaps with an existing path in
+          // the Docker image, which is very likely what would happen if we tried to mount the
+          // volume at Java's temporary directory path, which is /tmp in many JDKs.
+          val resolvedLocalDirs = sparkConf.get(
+              "spark.local.dir", s"$GENERATED_LOCAL_DIR_MOUNT_ROOT/${UUID.randomUUID()}")
+              .split(",")
+          val localDirVolumes = resolvedLocalDirs.zipWithIndex.map { case (dir, index) =>
+            new VolumeBuilder()
+              .withName(s"spark-local-dir-$index-${Paths.get(dir).getFileName.toString}")
+              .withNewEmptyDir().endEmptyDir()
+              .build()
+          }
+          val localDirVolumeMounts = localDirVolumes.zip(resolvedLocalDirs).map {
+            case (volume, path) =>
+              new VolumeMountBuilder()
+                .withName(volume.getName)
+                .withMountPath(path)
+                .build()
+          }
+          // Setting the SPARK_LOCAL_DIRS environment variable will force the executor to use the
+          // generated directory if the user did not provide one, as opposed to using the Java
+          // temporary directory. This also overrides the value of spark.local.dir in SparkConf,
+          // which is intended. See Utils#getConfiguredLocalDirs().
```

Comment:
why set `SPARK_LOCAL_DIRS` instead of `SPARK_EXECUTOR_DIRS` or letting that method fall through to read from `spark.local.dir` ?  Seems like we could not set any envvar and have the same result in the executor

Reply:
+1. Hope we can stick to one mechanism.

Reply:
We don't set `spark.local.dir` for the user on the driver side, which means that when the executor picks up the configuration, it will get whatever the driver has set for `spark.local.dir`. So we want to override that for the executors.

Reply:
I think the question boils down to whether or not we want the driver to also have emptyDir volume mounts for its local directories. This didn't seem necessary since the driver should not be doing any write-heavy operations.

Reply:
Good point. I wonder what's the K8s overhead of creating unused emptyDir for the driver. @foxish?

Reply:
I suspect the driver does have some operations where its block manager writes to local disk, like large `take` or `collect` calls.

If it's a good idea to put executors local dirs onto an EmptyDir, seems like it would be a good idea to put driver's local dirs onto an EmptyDir too for consistency

Reply:
There's a potential mismatch here though. In the case of using the external shuffle service, we won't be mounting the local directories as `emptyDir` volumes on the executors, but rather as `hostPath` volumes. But we would still mount them as `emptyDir` volumes for the driver. I suppose if it's reasonable to have inconsistent volume mounts between the driver and the executor then that's fine - this seems like the right semantics anyways.

Reply:
SGTM. It's already messy a bit and we're not making it worse.

Reply:
On the code level - we probably don't want to use a bootstrap module that's shared between the submission client and the driver, since the logic is different enough between the two.